### PR TITLE
[Inductor cutlass backend] Experimental extended Cutlass op generator

### PR DESCRIPTION
**Pull request by kadeng**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119701
* #119600
* __->__ #119009
* #119008
* #119601
* #119007
* #119598
* #119597
* #119006
* #119004
* #120620

So far we have relied on the "generator.py" from
third_party/cutlass/python/cutlass_library/ to generate Cutlass op
configurations to be tried in autotuning. These op configs are not
sufficient to ensure good performance and dtype coverage in all cases.

This diff introduces an extended variant of that generator.py which imports all symbols from generator.py overrides some functions from that file and reexports everything such that it can be used as a drop-in replacement for generator.py

Test Plan:
* CI
* python test/inductor/test_max_autotune.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by aakhundov** [comment]
@kadeng while I understand the motivation, I believe, this will be hard to keep in sync with the upstream CUTLASS generator. Few questions:

1. What's the scale of the change introduced in the `generator_extended.py` in this diff compared to the original generator in CUTLASS? Is the whole thing new, or most is copy/pasted and adapted in some small places?

2. What's the scale of the perf benefit the extended generator brings? Is this benefitting perf across the board or tailored for specific cases?

3. Instead of adding a whole new extended generator, can we post-process the ops that the original CUTLASS generator returns (e.g., change / filter / duplicate with new attribute values)?

4. The PR summary mentions the diff, but I don't see it in the changed files: there is only `generator_extended.py`. Where is the diff?

5. Is it feasible to upstream the changes in the extended generator to `nvidia/cutlass`? This way, we won't have to worry about keeping the local and upstream version in sync.

Thanks!

**Comment by kadeng** [comment]
> @kadeng while I understand the motivation, I believe, this will be hard to keep in sync with the upstream CUTLASS generator. Few questions:

I agree, that's also why I separated out this change and called it experimental, so we can discuss separately whether we want to keep it.

> 1. What's the scale of the change introduced in the `generator_extended.py` in this diff compared to the original generator in CUTLASS? Is the whole thing new, or most is copy/pasted and adapted in some small places?

It's mostly small changes. It will bring some slight performance improvements for smaller shapes, and these changes were part of the benchmarking in my WP note. A full ablation study has not been done, but the main reason these changes exist is to attempt to overcome Cutlass performance weaknesses ( compared to Triton ) for some shapes where at least one dimension is small  ( e.g. < 32 ). I made these changes after some suggestions by Cutlass devs.

> 2. What's the scale of the perf benefit the extended generator brings? Is this benefitting perf across the board or tailored for specific cases?

Specific cases, and it's minor. It's mostly to make Cutlass more competitive with Triton for some shapes, but these are usually covered well by Triton as well, so if Cutlass has an edge there, it is usually a small one, if at all.

> 3. Instead of adding a whole new extended generator, can we post-process the ops that the original CUTLASS generator returns (e.g., change / filter / duplicate with new attribute values)?

I would consider that less safe than doing it like this.

> 4. The PR summary mentions the diff, but I don't see it in the changed files: there is only `generator_extended.py`. Where is the diff?

That's a remnant from a previous code version. I will update the description. At the moment the generator_extended.py itself works like a runtime patch to cutlass_library.generator by importing all from that package, selectively overriding some functions. That's easier to maintain and read.

> 5. Is it feasible to upstream the changes in the extended generator to `nvidia/cutlass`? This way, we won't have to worry about keeping the local and upstream version in sync.

I agree, that's preferable. 

So, to summarize. I would agree that it can make sense to cut out these changes and instead try to get corresponding changes upstreamed in Cutlass itself. Agreed? If yes, I would remove this PR from the stack.

**Comment by aakhundov** [comment]
> So, to summarize. I would agree that it can make sense to cut out these changes and instead try to get corresponding changes upstreamed in Cutlass itself. Agreed? If yes, I would remove this PR from the stack.

Given the marginal improvements in special cases, I think, we should not land this and bear the maintenance burden in syncing with upstream CUTLASS changes. Upstreaming to nvidia/cutlass sounds more preferable.

**Comment by kadeng** [comment]
Closing. The PR stack is being restructured here: https://github.com/pytorch/pytorch/pull/121492
